### PR TITLE
Refacto unit tests

### LINKED ISSUE #103: Refactorize some unit tests common to MapieRegressor and MapieClassifier
**Is your feature request related to a problem? Please describe.**
Certain unit tests are copy-pasted from `test_regression.py` to `test_classification.py`.

**Describe the solution you'd like**
These tests could be refactorized in a `test_common.py` file, with `pytest.mark.parametrize` to repeat a piece of code across different instances of `MapieRegressor` and `MapieClassifer`

**Additional context**
These tests are `test_initialized`, `test_fit`, `test_fit_predict`, `test_no_fit_predict`,  `test_default_sample_weight`, `test_default_alpha`, `test_invalid_estimator`, `test_sklearn_compatible_estimator`.

Others tests can probably be factorized easily this way.

### PR DESCRIPTION
**Pull request by gmartinonQM**
# Description

* Refactorize some unit tests with code common between `test_regression.py` and `test_classification.py` into a `test_common.py` file with pytest parametrization.
* Realign unit tests in the same order between `test_regression.py` and `test_classification.py`
* Also simplifies github issues templates
* Fix README
* Fix all documentation bugs
* Increase dependencies versions (needed to fix documentation bugs"
* Simplify CI

Fixes #103 #108 

## Type of change

Please remove options that are irrelevant.

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

With pytest, covarege still at 100%.

# Checklist:

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

File: mapie/tests/test_classification.py [hunk 1]
```diff
@@ -357,61 +316,47 @@ def test_invalid_prefit_estimator(estimator: ClassifierMixin) -> None:
...
-    mapie = MapieClassifier(estimator=model, cv=cv)
-    mapie.fit(X_toy, y_toy)
+    mapie_clf = MapieClassifier(method=method)
+    mapie_clf.fit(X_toy, y_toy)
+    check_is_fitted(mapie_clf, mapie_clf.fit_attributes)
 
 
 @pytest.mark.parametrize(
-    "cv", [-3.14, 1.5, -2, 0, 1, "cv", DummyClassifier(), [1, 2]]
+    "cv", [-3.14, -2, 0, 1, "cv", DummyClassifier(), [1, 2]]
```
Comment (vtaquet) [inline mapie/tests/test_classification.py:344]: Je garderais ce 1.5 comme valeur testée

File: mapie/tests/test_common.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    return [MapieRegressor, MapieClassifier]
+
+
+@pytest.mark.parametrize("MapieEstimator", MapieEstimators())
+def test_initialized(MapieEstimator: BaseEstimator) -> None:
+    """Test that initialization does not crash."""
+    MapieEstimator()
+
+
+@pytest.mark.parametrize("MapieEstimator", MapieEstimators())
```
Comment (vtaquet) [inline mapie/tests/test_common.py:29]: Placerait-on ici aussi le check des valeurs par défaut des arguments communs aux deux classes (pour le moment juste `n_jobs` et `verbose`) ? Dans un souhait de garder  les deux classes homogènes, cela permettrait d'expliciter quels sont les arguments communs aux deux classes et ceux spécifiques.

File: mapie/tests/test_regression.py [hunk 3]
```diff
@@ -200,20 +153,7 @@ def test_valid_prefit_estimator(estimator: RegressorMixin) -> None:
     estimator.fit(X_toy, y_toy)
```
Comment (vtaquet) [inline mapie/tests/test_regression.py:153]: Fonction à déplacer dans `test_commons` ?

File: mapie/tests/test_regression.py [hunk 4]
```diff
@@ -123,53 +116,13 @@ def test_default_parameters() -> None:
...
-
-
-def test_no_fit_predict() -> None:
-    """Test that predict before fit raises errors"""
-    mapie_reg = MapieRegressor(estimator=DummyRegressor())
-    with pytest.raises(NotFittedError, match=r".*not fitted.*"):
-        mapie_reg.predict(X_toy)
-
-
 def test_none_estimator() -> None:
```
Comment (vtaquet) [inline mapie/tests/test_regression.py:119]: Fonction à déplacer également dans `test_commons` avec des Tuple `(MapieRegressor, LinearRegression)` et `(MapieClassifier, LogisticRegression)` ?

File: mapie/tests/test_regression.py [hunk 5]
```diff
@@ -464,6 +329,67 @@ def test_results_with_constant_sample_weights(strategy: str) -> None:
...
+
+
+@pytest.mark.parametrize("method", ["plus", "minmax"])
+@pytest.mark.parametrize("cv", [-1, 2, 3, 5])
+@pytest.mark.parametrize("alpha", [0.05, 0.1, 0.2])
+def test_prediction_agg_function(
+    method: str, cv: Union[LeaveOneOut, KFold], alpha: int
+) -> None:
+    """
+    Test that predictions differ when agg_function if None/"mean",
```
Comment (vtaquet) [inline mapie/tests/test_regression.py:349]: when agg_function is None/"mean"

File: mapie/tests/test_common.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    estimator: Any
+) -> None:
+    """Test that invalid estimators raise errors."""
+    mapie = MapieEstimator(estimator=estimator)
+    with pytest.raises(ValueError, match=r".*Invalid estimator.*"):
+        mapie.fit(X_toy, y_toy)
+
+
+@parametrize_with_checks([MapieRegressor()])  # type: ignore
+def test_sklearn_compatible_estimator(
```
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/tests/test_common.py:79]: Oui, pour l'instant il y a juste MapieRegressor dans la liste, mais j'ai mis dans la TODO d'ajouter MapieClassifier dès que possible.

File: mapie/tests/test_classification.py (hunk 1 above)
Comment (gmartinonQM) [reply mapie/tests/test_classification.py:344]: Done

File: mapie/tests/test_regression.py (hunk 5 above)
Comment (gmartinonQM) [reply mapie/tests/test_regression.py:349]: Done

File: mapie/tests/test_common.py (hunk 2 above)
Comment (gmartinonQM) [reply mapie/tests/test_common.py:29]: Done

File: mapie/tests/test_regression.py (hunk 4 above)
Comment (gmartinonQM) [reply mapie/tests/test_regression.py:119]: Done

File: mapie/tests/test_regression.py (hunk 3 above)
Comment (gmartinonQM) [reply mapie/tests/test_regression.py:153]: Done

File: doc/conf.py [hunk 7]
```diff
@@ -71,7 +71,7 @@
 # source_encoding = "utf-8-sig"
 
 # Generate the plots for the gallery
-plot_gallery = "True"
+plot_gallery = True
```
Comment (vtaquet) [inline doc/conf.py:74]: Does it change anything in practice ? The gallery was already generated.

File: doc/tutorial_regression.rst [hunk 8]
```diff
@@ -1,5 +1,5 @@
-Tutorial for regression
-=======================
+Tutorial
+========
```
Comment (vtaquet) [inline doc/tutorial_regression.rst:2]: The tutorial is now automatically generated from a notebook, I will add the .md file in PR #119

File: mapie/regression.py [hunk 9]
```diff
@@ -68,9 +68,9 @@ class MapieRegressor(BaseEstimator, RegressorMixin):  # type: ignore
           ``sklearn.model_selection.LeaveOneOut()``.
         - CV splitter: any ``sklearn.model_selection.BaseCrossValidator``
           Main variants are:
-            - ``sklearn.model_selection.LeaveOneOut`` (jackknife),
-            - ``sklearn.model_selection.KFold`` (cross-validation),
-            - ``subsample.Subsample`` object (bootstrap).
+          - ``sklearn.model_selection.LeaveOneOut`` (jackknife),
```
Comment (vtaquet) [inline mapie/regression.py:71]: Should we add blank lines between bullet points also here ?

File: mapie/subsample.py [hunk 10]
```diff
@@ -43,7 +43,7 @@ class Subsample(BaseCrossValidator):  # type: ignore
 
     def __init__(
         self,
-        n_resamplings: int,
+        n_resamplings: int = 30,
```
Comment (vtaquet) [inline mapie/subsample.py:46]: Add default value in docstring.

File: mapie/tests/test_classification.py [hunk 11]
```diff
@@ -334,229 +332,36 @@ def test_initialized() -> None:
...
-@pytest.mark.parametrize("estimator", [0, "estimator", KFold(), ["a", "b"]])
-def test_invalid_estimator(estimator: Any) -> None:
-    """Test that invalid estimators raise errors."""
-    mapie = MapieClassifier(estimator=estimator)
-    with pytest.raises(ValueError, match=r".*Invalid estimator.*"):
-        mapie.fit(X_toy, y_toy)
+    mapie_clf = MapieClassifier()
+    assert mapie_clf.method == "score"
+    assert mapie_clf.cv == "prefit"
+    assert mapie_clf.verbose == 0
```
Comment (vtaquet) [inline mapie/tests/test_classification.py:338]: To be removed as this check is in `test_common.py`.

File: mapie/tests/test_classification.py [hunk 12]
```diff
@@ -871,3 +660,90 @@ def test_classifier_without_classes_attribute(
         AttributeError, match=r".*does not contain 'classes_'.*"
     ):
         mapie.fit(X_toy, y_toy)
+
+
+@pytest.mark.parametrize("X_wrong_image", X_WRONG_IMAGE)
+def test_wrong_image_shape_fit(X_wrong_image: ArrayLike) -> None:
+    """
+    Test that VaueError is raised if image has not 3 or 4 dimensions in fit.
```
Comment (vtaquet) [inline mapie/tests/test_classification.py:668]: Typo ValueError

File: mapie/tests/test_classification.py [hunk 13]
```diff
@@ -871,3 +660,90 @@ def test_classifier_without_classes_attribute(
...
+        random_state=42
+    )
+    with pytest.raises(ValueError, match=r"Invalid X.*"):
+        mapie.fit(cumclf.X_calib, cumclf.y_calib, image_input=True)
+
+
+@pytest.mark.parametrize("X_wrong_image", X_WRONG_IMAGE)
+def test_wrong_image_shape_predict(X_wrong_image: ArrayLike) -> None:
+    """
+    Test that VaueError is raised if image has not
```
Comment (vtaquet) [inline mapie/tests/test_classification.py:685]: Typo ValueError

**Review by vtaquet — CHANGES_REQUESTED** [review]
Still some minor changes to do before merging.

File: doc/conf.py (hunk 7 above)
Comment (gmartinonQM) [reply doc/conf.py:74]: No, but it removes a warning when building the doc with sphinx

File: doc/tutorial_regression.rst (hunk 8 above)
Comment (gmartinonQM) [reply doc/tutorial_regression.rst:2]: OK

File: mapie/regression.py (hunk 9 above)
Comment (gmartinonQM) [reply mapie/regression.py:71]: No, these are sub-bullet points, an sphinx does not complain about them

File: mapie/subsample.py (hunk 10 above)
Comment (gmartinonQM) [reply mapie/subsample.py:46]: Done

File: mapie/tests/test_classification.py (hunk 11 above)
Comment (gmartinonQM) [reply mapie/tests/test_classification.py:338]: Done

File: mapie/tests/test_classification.py (hunk 12 above)
Comment (gmartinonQM) [reply mapie/tests/test_classification.py:668]: Done

File: mapie/tests/test_classification.py (hunk 13 above)
Comment (gmartinonQM) [reply mapie/tests/test_classification.py:685]: Done

**Pull request merged**